### ISSUE TITLE
[Master Feature] Allow ads to be position fixed on desktop canoncial

### ISSUE THREAD

**Issue by bjalford** [ampproject/amphtml#18123]
We have a requirement to make an ad sticky in an article template for a publisher in the right hand column on desktop (AMP as canonical). If we add a position fixed the ad doesn't render.

In the same way the amp-sticky-ad has a height restriction, could amp-ads on desktop (or media query great than x pixels) that are position fixed have a max width. Our specific requirement is to render MPUs and double MPUs (300 wide).

**Comment by zhouyx** [comment]
to @jasti Regarding how sticky ad works on desktop. 

@bjalford that's right, AMP won't render an ad if it is position fixed unless it is positioned inside flying-carpet/lightbox/sticky-ad container. 
The request here is to have the sticky-ad sticky to the side of the page (rather than bottom) on desktop? And then maybe a max width restriction can be applied.

**Comment by bjalford** [comment]
@zhouyx correct. My exact requirements would be for it to be inside a sidebar container (max-width) and be the bottom item inside that container which would then stick.

Example can be seen here: https://www.independent.co.uk/news/uk/home-news/northern-ireland-abortion-pill-mother-prosecuted-daughter-belfast-court-a8546936.html
- Under 'Popular Videos' an ad (JS DFP) is rendering which we then stick as the user scrolls down the page

**Comment by jasti** [comment]
@bjalford unfortunately large sticky ads on desktop are not approved by the better ads standard as a good ad/ user experience, so we have had to limit the height of the sticky ad on both desktop & mobile. https://www.betterads.org/desktop-large-sticky-ad/

**Comment by zhouyx** [comment]
@bjalford As Vamsee said. Large sticky ads on desktop is not something we are planning to build for better ad/user experience. Close as not intent to implement.

**Comment by jasti** [comment]
Reopening this after more research. We'll look into supporting this since
a) This is referring to the sticky ad being in the right or left rail in a fixed position in the desktop view 
b) It doesn't obstruct content and is not simply a larger version of the current sticky ad on mobile

**Comment by bcoonsleg** [comment]
We would like to use this feature and were disappointed to find it was marked Stale.  Please bring this back to life.

**Comment by mrjjones** [comment]
Why was this closed?   This seems like a key feature to have in AMP.

**Comment by zhouyx** [comment]
Added @lannka and @powerivq 
Would the new sticky format proposal help here?

**Comment by adekbadek** [comment]
Looks like the feature is proposed in more detail here: https://github.com/ampproject/amphtml/issues/31513

**Comment by adamsilverstein** [comment]
@adekbadek - The feature discussed in #31513 is more about the top sticky on mobile unit, where a thin ad unit sticks across the top of the mobile viewport, right? 

This issue and [discussion](https://github.com/ampproject/amphtml/issues/18123#issuecomment-429694809) are about an ad in the side bar/rail that sticks to the top as the user scrolls down the page. Generally this can be achieved by placing the ad unit in a container with `position:fixed` or `position:sticky` applied, which amp-ad explicitly disallowed.  

To enable this in AMP we would need to ensure it was only available on desktop, and limit the width of the fixed/sticky container - so it should only work:
- for ads up to a certain width, such as a sidebar 300x250 unit, and
- for screen sizes above a certain minimum size, enabling only side ads that take a small percentage of the total width.

Any other details to consider here?

**Comment by adekbadek** [comment]
> The feature discussed in #31513 is more about the top sticky on mobile unit, where a thin ad unit sticks across the top of the mobile viewport, right?

Yes, looks like it.

**Comment by pradeep910** [comment]
@adamsilverstein 

I think we can add `sticky="left"` and `sticky="right"` attribute to achieve it. 

Since https://github.com/ampproject/amphtml/issues/31513 already has support for top sticky ad, we can keep `left` and `right` values just for desktop devices and limit sizes in CSS. 

This would essentially be used for skyscraper ads on desktop such as:
- 160x600
-  125x600
-  300x600
-  and probably 300x250

**Comment by claudiulodro** [comment]
I'm curious, with this approach will it be possible to do sticky sidebar ad on desktop viewports **and** sticky footer ad on mobile viewports? Currently there's a one sticky-ad-per-page restriction in AMP, so it would be ideal if sites didn't need to choose between sticky desktop sidebar **or** sticky mobile footer ads.

**Comment by pradeep910** [comment]
@claudiulodro Yes for both. 

I think there are 2 types of ads:
1. **Sticky sidebar ads** - uses `position:sticky` and sticks to top only when it appears in viewport and if user scrolls further to the bottom. Here the sticky sidebar ad is part of the 2 or 3 column layout.
2. **Gutter ads** - uses `position:fixed` and sticks to both sides of content box. (see the example screenshot below.) 

> **_Currently there's a one sticky-ad-per-page restriction in AMP_**

I think we need to do some changes to how ads are being loaded on desktop in general. The desktop can show both gutter ads and footer sticky ads or maybe either of them if a site owner wants to show them.
The sticky ads may not be ideal for all types of websites, so it depends on the site owners actually. Sticky ads guidelines for reference - https://support.google.com/admanager/answer/7246067?hl=en

### For Desktop Viewports

On desktop devices, the websites mostly have space left for the "gutter ads" (vertical sticky ads on both sides of a page), so sites like [Loksatta.com](https://www.loksatta.com/), [IndianExpress.com](https://indianexpress.com/) show gutter ads. 
We may have to remove some restrictions for desktop only, such as this piece of code - https://github.com/ampproject/amphtml/blob/a0ba91919772989e226ad40bb5b12047c920f613/extensions/amp-a4a/0.1/amp-a4a.js#L639-L646

This will enable us to show `position:fixed` gutter ads on desktop. _(currently we get warning and AMP doesn't show such vertical sticky `amp-ad` on desktop)_

I see most news websites keep such gutter ads (sticky ads) on their websites, so I think we can also add this gutter ads option along side the "Enable global header ad" option as an enhancement to [newspack-ads](https://github.com/Automattic/newspack-ads).

The "gutter ads" are highlighted part in the screenshot below.

![Screenshot 2021-07-08 at 10 40 02 PM](https://user-images.githubusercontent.com/1139853/124964376-94902180-e03e-11eb-82a6-55e090c14f2c.png)

### For mobile viewports
We can continue to use 1 sticky ad (top or bottom). The gutter ads shown above won't be shown on mobile anyway.

**Comment by pradeep910** [comment]
Here is the graphic representation of the different sticky ads I was referring to. 

| Ad type  | Ad positions  |
|---|---|
|  Desktop: sticky gutter ads + sticky footer ad |  ![ezgif com-gif-maker (2)](https://user-images.githubusercontent.com/1139853/125077096-53e7e500-e0de-11eb-84da-e73fbe991eb6.gif) |
|  Desktop: only sticky gutter ads | ![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/1139853/125077125-5cd8b680-e0de-11eb-860b-41de0d9abb18.gif)|
|  Desktop: sticky sidebar ad  | ![ezgif com-gif-maker](https://user-images.githubusercontent.com/1139853/125077135-6104d400-e0de-11eb-89f9-f51c38e42a99.gif) |

**Comment by adamsilverstein** [comment]
@pradeep910 These graphics are super helpful for understanding the units, thank you. This is a good starting point for getting feedback on adding support for these types of units to AMP.

In all these cases the ad is fixed (inside a container with `position: fixed`). The sticky sidebar ads I was referring to look a little different - they scroll into view in the sidebar as the user scrolls down the page, then stick at their container top. When another unit scrolls into position the first unsticks and scrolls out. This is achieved with `position:sticky` on the container. I'd like to consider this unit as well because it is less obtrusive than the fixed units and still has great view-ability.

Here is a video I captured showing what this looks like on a live site:


https://user-images.githubusercontent.com/2676022/125117100-ebcbdb80-e0aa-11eb-9793-3aa30cc22eea.mp4

**Comment by kristoferbaxter** [comment]
@lannka or @powerivq can you take a look?

**Comment by powerivq** [comment]
We left open the possibility to do this when designing the new `<amp-ad sticky>` sticky ads. It looks feasible. Questions are:
1. How do we ensure the gutter ads do no encroach on the content space? Is it up to publisher to place an appropriate media query to block them for mobile pages?
2. It looks like a change on the size and numerical limit must be made. How large are we comfortable to let theses ads be? @nainar Or maybe as a starting point, how does the proposed size above sound?

**Comment by pradeep910** [comment]
To ensure the gutter ads do not overlap on content we can add media query
so the gutter ads display only on desktop. Also, to some extent it would be
on site owners or publishers to make sure they don't put it on full screen
content pages or box content with very little margin.

As long as the content atleast has 200px available on both sides we would
be good to go with 125x600 and 160x600 sizes.

What do you think?

**Comment by pradeep910** [comment]
https://github.com/ampproject/amphtml/blob/main/src/ad-helper.js#L30

We experimented further on this and currently we can see it working with `<amp-sticky-ad>` component. 

If you look at this, only these 4 components or containers are allowed to use sticky or fixed positions. 
https://github.com/ampproject/amphtml/blob/a0ba91919772989e226ad40bb5b12047c920f613/src/ad-helper.js#L27-L32

There are two possible options we go with:

1. **Option 1**: Add `<amp-ad>` in the containers list to allow use of `position:sticky` or `position:fixed` in such a way that doesn't affect current `<amp-sticky-ad>` that we use for sticky footer ad.

2. **Option 2**: Use `<amp-sticky-ad>` component as parent of `<amp-ad>` for gutter or sidebar sticky ads. 

Have a look at the following demos which are using `<amp-sticky-ad>` component as parent of `<amp-ad>` for both sticky gutter ads and sticky sidebar ads. This would be inline with current sticky footer ads IMO. What are your thoughts? 

I think @AnuragVasanwala is working on this further to figure out the best way in the meantime.

### Sticky Gutter Ads
https://user-images.githubusercontent.com/1139853/125483130-fae5dd2a-02c7-4a0d-9f6a-1c8a5b6a0536.mp4

### Sticky Sidebar Ads
https://user-images.githubusercontent.com/1139853/125483156-6b55f312-f45a-4af0-939c-38870342309a.mp4

cc @adamsilverstein @kristoferbaxter @powerivq @nainar

**Comment by powerivq** [comment]
@pradeep910 We are not inclined to allow customization by arbitrary CSS rules. We currently have `<amp-ad sticky=bottom xxx>` and `sticky=top` components. The way I think we will do is to extend it to allow `sticky=left` and `sticky=right`.

**Comment by alanorozco** [comment]
cc @ampproject/wg-components

**Comment by powerivq** [comment]
@zewenliang

**Comment by powerivq** [comment]
As a starting point, we propose starting allowing this for viewport of at least 1024 pixels wide, and to allow multiple sticky ads on one page (only for screens >1024 wide). We can revisit the limit after launching. @pradeep910 

@zewenliang WDYT?

**Comment by powerivq** [comment]
@AnuragVasanwala Here is our plan for these ad formats.

**Comment by AnuragVasanwala** [comment]
> As a starting point, we propose starting allowing this for viewport of at least 1024 pixels wide, and to allow multiple sticky ads on one page (only for screens >1024 wide). We can revisit the limit after launching. @pradeep910
> 
> @zewenliang WDYT?

Sounds good. I have added CSS size restriction for `gutter-ad` into [🧪 Experimental PR](https://github.com/ampproject/amphtml/pull/35216/commits/f08872a43ca85d9d9df8b5cf68d82c99e724c800). I am soon planning to add size restriction into `amp-ad` component itself so that it won't render on devices having viewport less than `1024` pixels wide.

**Comment by adamsilverstein** [comment]
@AnuragVasanwala Checking in on the status here, are you still waiting for feedback on the PR(s)?

cc: @zewenliang, @powerivq Can you provide some feedback on the PRs here or do you have any questions on the approach?

**Comment by AnuragVasanwala** [comment]
@adamsilverstein Yes, actually I am waiting for the feedback on the PR https://github.com/ampproject/amphtml/pull/35216. So, I can have better idea what to correct and where to improve.

**Comment by powerivq** [comment]
@AnuragVasanwala I have introduced https://github.com/ampproject/amphtml/pull/36135 in the interest of simplicity.

**Comment by powerivq** [comment]
@westonruter It has been adjusted. Will be deployed soon.

**Comment by raja-anbazhagan** [comment]
@westonruter I can see that the validators are available in the recent release.

https://github.com/ampproject/amphtml/commit/78e8f7a9b45fbbe2cac2c338c88bcddb76ee59f5

**Comment by westonruter** [comment]
@raja-anbazhagan I've updated the spec in the plugin (https://github.com/ampproject/amp-wp/pull/6697). You can access the spec update in the latest 2.1.x build of the plugin via our [Development Builds](https://github.com/ampproject/amp-wp/wiki/Development-Builds) page. (See [diff](https://github.com/ampproject/amp-wp/compare/2.1.4...2.1) with current stable.)

**Comment by ahmednagarnews** [comment]
I want News Seo plugins for WordPress <a href="https://www.akolenews.com/sangamner-taluka-news/">Sangamner News </a>
